tests/h/services/list_groups_test.py

```diff
@@ -161,6 +161,32 @@ def test_groups_are_sorted(self, list_groups_service, factories):
         assert [group.pubid for group in groups] == ['zoinks', 'aaaa', 'zzzz']
 
 
+class TestListGroupsParseOrigin(object):
+
+    @pytest.mark.parametrize('document_uri', [
+        (u'http://www.foo.bar:80/ding', u'http://www.foo.bar:80'),
+        (u'http://www.foo.bar:80/', u'http://www.foo.bar:80'),
+        (u'http://www.foo.bar:80/flop.html', u'http://www.foo.bar:80'),
+        (u'http://www.foo.bar:80/flop.html#fragment', u'http://www.foo.bar:80'),
+        (u'https://foo.bar/', u'https://foo.bar'),
+        (u'https://userfoo:hitherepassword@foo.bar/zowie/bang.pdf', u'https://userfoo:hitherepassword@foo.bar'),
+        (u'//zounds.com', u'//zounds.com')
+    ])
```

Comment:
Ah, the first argument to `parametrize` (`'document_uri'`) is actually a comma-separated list of parameters to paremetrize the test by. So instead of `@pytest.mark.parametrize('document_uri', [` you can do `@pytest.mark.parametrize('document_uri,origin', [`, and then the test method can take two separate params `document_uri` and `origin`, instead of having to use `document_uri[0]` and `document_uri[1]`. I think it'll make the test more readable:

```diff
diff --git a/tests/h/services/list_groups_test.py b/tests/h/services/list_groups_test.py
index 20f38c840..0d558a63d 100644
--- a/tests/h/services/list_groups_test.py
+++ b/tests/h/services/list_groups_test.py
@@ -163,7 +163,7 @@ class TestListGroupsOpenGroups(object):
 
 class TestListGroupsParseOrigin(object):
 
-    @pytest.mark.parametrize('document_uri', [
+    @pytest.mark.parametrize('document_uri,origin', [
         (u'http://www.foo.bar:80/ding', u'http://www.foo.bar:80'),
         (u'http://www.foo.bar:80/', u'http://www.foo.bar:80'),
         (u'http://www.foo.bar:80/flop.html', u'http://www.foo.bar:80'),
@@ -172,10 +172,10 @@ class TestListGroupsParseOrigin(object):
         (u'https://userfoo:hitherepassword@foo.bar/zowie/bang.pdf', u'https://userfoo:hitherepassword@foo.bar'),
         (u'//zounds.com', u'//zounds.com')
     ])
-    def test_it_returns_origin_from_uri_string(self, list_groups_service, document_uri):
-        result = list_groups_service._parse_origin(document_uri[0])
+    def test_it_returns_origin_from_uri_string(self, list_groups_service, document_uri, origin):
+        result = list_groups_service._parse_origin(document_uri)
 
-        assert result == document_uri[1]
+        assert result == origin
 
     @pytest.mark.parametrize('document_uri', [
         (None, None),
```

Reply:
I wonder if `_parse_origin()` might be useful in a utility module somewhere, where any code could potentially import and use it? I could imagine it maybe coming in handy one day. But also `parse_origin()` would then be part of the public API so said utility module, so you could unit test it without having to unit test a `_private` method.

Reply:
> I wonder if _parse_origin() might be useful in a utility module somewhere, where any code could potentially import and use it? 

I like this idea, but am unsure where to stick it. The _intuitive_ place seems like `utils/uri`, but that module is wholly concerned with the notion of normalizing a URI per annotations. So I'm not sure? Thoughts?

This also reminds me that I had a question about the "OK-ness" of testing private methods directly; I did see it happening in other tests so...

Reply:
I would say `utils/uri` or, if you think that module is getting too big, we could turn it into `utils/uri/normalize.py` and `utils/uri/parse.py`.